### ISSUE TITLE
.tsv file extension support 

### ISSUE THREAD

**Issue by jzf2101** [jupyterlab/jupyterlab#4226]
See https://github.com/jupyterlab/jupyterlab/pull/4192 discussion @jasongrout

**Comment by jasongrout** [comment]
This is an 'advanced' good first issue - it will take a bit of work to refactor the classes to allow us to pass the delimiter into the constructor options, but I think it's probably fairly straightforward for someone that already has a background in programming or javascript/typescript.

**Comment by jasongrout** [comment]
I think the necessary refactoring is in #4453

**Comment by jzf2101** [comment]
Does that mean that we can close with #4453 ?

**Comment by jasongrout** [comment]
No, the .tsv is still not provided. Just that I touched the csv viewer code and I think introduced the changes needed to make a PR solving this issue easier.

**Comment by jasongrout** [comment]
(After #4453, I think this would be a standard good first issue, rather than an advanced good first issue.)

**Comment by jirivrany** [comment]
OK great :) My current observation is, that the recent version handles tabs as the separator in csv pretty well. If I create file separated with tabs, but with .csv extension, it is open correctly, there is only need to switch the delimiter. However the same file, with the .tsv extension opens in a standard text editor.

If my observations are correct, then the csvviewer is created by CSVViewerFactory. And the registration of the factory to the extension is done in [the activation function of csvviewer-extension module](https://github.com/jupyterlab/jupyterlab/blob/master/packages/csvviewer-extension/src/index.ts). Am I on the right path?

**Comment by jasongrout** [comment]
Yes.

Probably the easiest thing to do is to create a new factory for tab-separated widgets, by copying and modifying https://github.com/jupyterlab/jupyterlab/blob/57845b9d6b406e0a992536bb990d00d16b8d63ed/packages/csvviewer/src/widget.ts#L209-L220, adding in the appropriate delimiter argument.

Then use that TSVDocumentFactory to set things up for tsv files similar to how things are set up for csv files in the extension you linked to above.

**Comment by jirivrany** [comment]
@jasongrout thank you. Just one more coding style related question. Should this new factory be in a new file, or it can be a part of widget.ts?

**Comment by jirivrany** [comment]
Yes, you are right, I was reading the code wrong. So I created the TSVFactory. I also registered tsv as a new file type and created two plugins. I took inspiration in the [tooltip extension package](https://github.com/jupyterlab/jupyterlab/blob/master/packages/tooltip-extension/src/index.ts). 

Unfortunately, I'm a bit stuck right now. Both plugins are working correctly if I register only one of them. However, if I try to export both of them in the list, only the first one is working correctly, the second not. [The code is in my repository](https://github.com/jirivrany/jupyterlab/blob/tsv-file-support/packages/csvviewer-extension/src/index.ts). I don't know where is the mistake. Can someone help, please?

**Comment by jasongrout** [comment]
I would just have one plugin and one activate function. Inside the one activate function, register both the csv and tsv factories and the tsv file type.

**Comment by jasongrout** [comment]
I was thinking in the future we could rename the package to `dsvviewer` for delimiter-separated viewer, rather than having the extension be only about commas. But I think that can be a separate PR.

**Comment by jirivrany** [comment]
Thanks @jasongrout. If you don't mind, I think that two code with two separate functions is cleaner. So I did this way. The issue was in the factory name, used in the restorer. I was using the same name twice, and that was the mistake. So I'm done, the PR was created. 

Some parts of the package are already using DSV terminology - model for example. So It is probably logical choice, to rename whole package.
